src/mac/frame/mod.rs

```diff
@@ -0,0 +1,993 @@
+
+        let taken_bytes = orginal_buf.len() - buf.remaining();
+        let payload_with_footer = &orginal_buf[taken_bytes..];
+
+        if payload_with_footer.len() != buf.remaining() {
+            panic!(
+                "Noncontinuous Buf implementation aren't supported. Consider use `bytes::Bytes`"
+            );
+        }
+        buf.advance(buf.remaining());
+
+        let mut footer = [0; 2];
+        let payload = if contains_footer {
+            if payload_with_footer.len() < 2 {
+                return Err(DecodeError::NotEnoughBytes);
+            }
+            let footer_pos = payload_with_footer.len() - 2;
+            footer.copy_from_slice(&payload_with_footer[footer_pos..]);
+            &payload_with_footer[..footer_pos]
+        } else {
+            payload_with_footer
+        };
+
+        Ok(Frame {
+            header,
+            content,
+            payload,
+            footer,
+        })
+    }
+
+    /// Encodes the frame into a buffer
+    ///
+    /// # Example
+    ///
+    /// ``` rust
+    /// use ieee802154::mac::{
+    ///     Address,
+    ///     ShortAddress,
+    ///     Frame,
+    ///     FrameContent,
+    ///     FrameType,
+    ///     FrameVersion,
+    ///     Header,
+    ///     PanId,
+    ///     WriteFooter,
+    ///     FrameControl,
+    ///     AddressMode,
+    /// };
+    ///
+    /// let frame = Frame {
+    ///     header: Header {
+    ///         frame_control: FrameControl {
+    ///         frame_type:      FrameType::Data,
+    ///         security:        false,
+    ///         frame_pending:   false,
+    ///         ack_request:     false,
+    ///         pan_id_compress: false,
+    ///         version:         FrameVersion::Ieee802154_2006,
+    ///         dest_addr_mode: AddressMode::Short,
+    ///         src_addr_mode: AddressMode::Short,
+    ///         },
+    ///         seq:             0x00,
+    ///
+    ///         destination: Some(Address::Short(PanId(0x1234), ShortAddress(0x5678))),
+    ///         source:      Some(Address::Short(PanId(0x1234), ShortAddress(0x9abc))),
+    ///     },
+    ///     content: FrameContent::Data,
+    ///     payload: &[0xde, 0xf0],
+    ///     footer:  [0x12, 0x34]
+    /// };
+    ///
+    /// let mut bytes = bytes::BytesMut::with_capacity(32);
+    ///
+    /// frame.encode(&mut bytes, WriteFooter::No);
+    /// let encoded_bytes = bytes.split().freeze();
+    ///
+    /// let expected_bytes = [
+    ///     0x01, 0x98,             // frame control
+    ///     0x00,                   // sequence number
+    ///     0x34, 0x12, 0x78, 0x56, // PAN identifier and address of destination
+    ///     0x34, 0x12, 0xbc, 0x9a, // PAN identifier and address of source
+    ///     0xde, 0xf0,             // payload
+    ///    // footer, not written
+    /// ];
+    /// assert_eq!(encoded_bytes[..], expected_bytes[..]);
+    /// ```
+    pub fn encode(&self, buf: &mut dyn BufMut, write_footer: WriteFooter) {
```

Comment:
In my opinion we should not enhance user to use raw `&mut [u8]` because  it can panic - (so it's not so rusty). I would prefer to wrap `&mut [u8]` in some wrapper and implement with `BufMut`:
- flag that validate slice end was reached
- allow determinate proper lengths (encoded/free)

I thought such crate should already exist - but could not find one. I could implement this and add it to example.
The second advantage with this implementation will be speed - if you are 100% sure you have enough space in slice - you don't need calculate byte counter. (I assume we want to return how many bytes was encoded from all `encode()` functions not just `Frame::encode()`)  This crate handle mac frames - so this can matter. 

If you need to call it often you can always create a helper function. 
```
fn encode_n(frame: &Frame, mut slice &mut[u8], write_fotter: WriteFooter) -> usize {
  let len_before = slice.len();
  frame.encode(&mut slice);
  return len_before - slice.len();
}
```

Reply:
The reason for my comment above is that I ported my [downstream crate](https://github.com/braun-embedded/rust-dw1000/tree/master/dw1000) to your branch and the code that calls `encode`, which had been straight-forward before, not only became more complicated, but it became hard to figure out what needed to be done in the first place.

I had to study `BufMut`, including its code, to figure out that it updates the slice in-place, and I can use `slice.len()` to figure out how much was written. This was fine for me (I had to understand `BufMut` anyway, for this review), but I think it puts too much of a burden on the user, especially since that was so easy before this change.

I think there are two solutions:
1. Just return the length, as I suggested. You make a good point about the panicking, but I'd be willing to live with that, as that's pretty much the same as we had previously.
2. We add a wrapper, as you suggested. I think this would be the better solution, and if you're willing to do that, that's great! If you don't have the time or inclination, I think solution 1. is a good-enough stopgap.

> (I assume we want to return how many bytes was encoded from all `encode()` functions not just `Frame::encode()`)

My downstream crate only uses the top-level `encode()`, which I assume is a fairly typical use of this crate. So this one is the most important. But handling it consistently for all `encode()` methods would be the best solution.

Reply:
I don't see ability to adding returning length just for `Frame::encode()` or I don't know how to do it. So change has to be done to all `encode()` what I would be doing anyway -- interface should be  the same for all functions.

> I had to study BufMut, including its code, to figure out that it updates the slice in-place, and I can use slice.len() to figure out how much was written. This was fine for me (I had to understand BufMut anyway, for this review), but I think it puts too much of a burden on the user, especially since that was so easy before this change.

I think we want to resolve #9. This is proposed cost of this - you need to study why `&mut [u8]` panics and should workaround. We can of course show in an example how to do it "proper".

I have found that there is [heapless](https://crates.io/crates/heapless) crate. Maybe it would be good way to add there PR that support `bytes::BufMut`.

I will have time to work on this after 24.08.

Reply:
Sounds good!

Reply:
Could you please test if [static-bytes](https://github.com/xoac/static-bytes) crate would be more user-friendly than `&mut [u8]` slice in your case?

I have created [one test](https://github.com/xoac/static-bytes/blob/290c1c366f94c71c1b75cad98d5a0981bb477a74/src/bytes_slice.rs#L104) that compares the `&mut [u8]` and `SafeBytesSlice`. It can be a good place to see how to use SafeBytesSlice